tools/src/bin/scip-indexer.rs

```diff
@@ -926,21 +944,72 @@ fn analyze_using_scip(
+            // not actually an official SCIP symbol, but because the local symbols
+            // are not namespaced, we just store it using our normalized version.
+            let (is_local, norm_scip_sym) = if occurrence.symbol.starts_with("local ") {
+                (
+                    true,
+                    format!(
+                        "S_{}_{}_{}/#{}",
+                        lang_name,
+                        subtree_name,
+                        sanitize_symbol(&doc.relative_path),
+                        &occurrence.symbol[6..]
+                    ),
+                )
+            } else {
+                (false, occurrence.symbol.clone())
+            };
+
+            let sinfo = match scip_symbol_to_structured.get(&norm_scip_sym) {
                 Some(s) => s,
                 None => {
-                    warn!(
-                        "Unable to find structured data for symbol: {}",
-                        occurrence.symbol
-                    );
-                    continue;
+                    if is_local {
+                        // For locals that don't exist, we create a new structured
+                        // fake, save it, and return it.
+
+                        let norm_scip_sym_ustr = ustr(&norm_scip_sym);
+                        let fake = AnalysisStructured {
+                            structured: StructuredTag::Structured,
+                            // XXX we should really try and extract the underlying token as the pretty.
+                            pretty: ustr(&occurrence.symbol),
+                            sym: norm_scip_sym_ustr,
+                            type_pretty: None,
+                            kind: ustr("local"),
+                            parent_sym: None,
+                            slot_owner: None,
+                            impl_kind: ustr("impl"),
+                            size_bytes: None,
+                            binding_slots: vec![],
+                            supers: vec![],
+                            methods: vec![],
+                            fields: vec![],
+                            overrides: vec![],
+                            props: vec![],
+
+                            idl_sym: None,
+                            subclass_syms: vec![],
+                            overridden_by_syms: vec![],
+                            extra: Map::default(),
+
+                        };
+                        scip_symbol_to_structured.insert(norm_scip_sym.clone(), fake);
+                        our_symbol_to_scip_sym.insert(norm_scip_sym_ustr, norm_scip_sym.clone());
+                        scip_symbol_to_structured.get(&norm_scip_sym).unwrap()
+                    } else {
+                        warn!(
+                            "Unable to find structured data for symbol: {}",
+                            norm_scip_sym
+                        );
+                        continue;
+                    }
```

Comment:
> At least for scip-typescript, it's very common to not have symbol defs for
> locals.  We now create synthetic structured records for this case that aren't
> really useful at all, but at least provide hover highlighting that's okay.

I also see this happen a lot with scip-java, not only for locals, but also for packages and types imported from dependencies.
For instance [this file](https://github.com/nicolas-guichard/mozsearch/blob/scip-java/tests/tests/files/java_project/lib/src/main/java/java_project/Library.java) and [this file](https://github.com/nicolas-guichard/mozsearch/blob/scip-java/tests/tests/files/java_project/lib/src/test/java/java_project/LibraryTest.java) are both part of the java_project package, and declare that with a `package java_project;`. scip-java lists both of those as occurrences of the symbol `semanticdb maven . . java_project/` but never actually provides a matching `SymbolInformation`.
In the [test file](https://github.com/nicolas-guichard/mozsearch/blob/scip-java/tests/tests/files/java_project/lib/src/test/java/java_project/LibraryTest.java) , this also applies to all external symbols (`org/`, `org/junit/`, ..., `org/junit/jupiter/api/Assertions#assertTrue(+4).`.

Would it be an issue to provide fake symbols in every cases, and not only for locals?

Reply:
> Would it be an issue to provide fake symbols in every cases, and not only for locals?

Not an issue.  We should definitely be emitting source and target records in these cases and my choice to have us `continue` here I think was probably a bit stop-gap; I expect if I'd tried to continue to improve the JS use-case that I would have changed this.  I believe I may have been under the (mistaken) impression that the SCIP file should be providing definitions in all cases, and this likely informed my attempt to assume we should always have a "structured" record.  (And thereby avoid me needing to handle the case where a structured record would not exist.)

It's even okay if we emit structured records that we infer in the analysis files that we produce.  To ensure we only emit the record once and are able to merge/de-duplicate it, we have only emitted the structured records at the type's point of definition.  The scip-indexer has things easier than `MozsearchIndexer.cpp` that operates in parallel, though, so it could arbitrarily pick a file to put structured records in as long as it can be consistent about the choice of file and "loc" so that `merge_files` in `merger.rs` can merge things consistently.  Maybe we could introduce a synthetic file like `__package__` or similar in order to hold inferred structured records whose scope is not inherently limited to a single file.  (I could believe we might end up wanting to emit structured records for the local symbols too.)

The Java "package" situation seems quite analogous to the C++ namespace situation where there's no single point of definition.  However, as I look at our handling of C++ namespaces right now, I see we [emit the namespaces as a "def"](https://searchfox.org/mozilla-central/rev/a1827d62f97d9721993ba489f4af7dea422d418f/build/clang-plugin/mozsearch-plugin/MozsearchIndexer.cpp#1670-1671) which feels wrong.  We should probably weaken that to a "decl" as a first step and maybe even introduce a specific kind like I did with "forward" for C++ forwards.  That said, there's no urgency to change the C++ namespaces to "decl" right now since it doesn't really affect any user functionality.